sw/device/lib/otbn.h

```diff
@@ -0,0 +1,248 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_otbn.h"
+
+/**
+ * @file
+ * @brief OpenTitan Big Number Accelerator (OTBN) driver
+ */
+
+/**
+ * Information about an embedded OTBN application image.
+ *
+ * All pointers reference data in the normal CPU address space.
+ *
+ * Use `OTBN_USE_APP()` together with `OTBN_APP_T_INIT()` to initialize this
+ * structure.
+ */
+typedef struct otbn_app {
+  /**
+   * Start of OTBN instruction memory.
+   */
+  const char *imem_start;
+  /**
+   * End of OTBN instruction memory.
+   */
+  const char *imem_end;
+  /**
+   * Start of OTBN data memory.
+   */
+  const char *dmem_start;
+  /**
+   * End of OTBN data memory.
+   */
+  const char *dmem_end;
+} otbn_app_t;
+
+/**
+ * A pointer to a symbol in OTBN's instruction or data memory.
+ *
+ * Use `OTBN_USE_PTR()` together with `OTBN_PTR_T_INIT()` to initialize this
+ * structure.
+ */
+typedef struct otbn_ptr {
+  /**
+   * Address of the symbol in normal CPU address space.
+   */
+  const char *ptr_in_cpu_addr_space;
+} otbn_ptr_t;
+
+/**
+ * The result of an OTBN operation.
+ */
+typedef enum otbn_result {
+  /**
+   * The operation succeeded.
+   */
+  kOtbnOk = 0,
+  /**
+   * An unspecified failure occurred.
+   */
+  kOtbnError = 1,
+  /**
+   * The precondition check for an argument passed into the function failed.
+   */
+  kOtbnBadArg = 2,
+  /**
+   * The chosen application is not part of the loaded application.
+   */
+  kOtbnFuncNotInApp = 3,
+  /**
+   * No application is loaded.
+   */
+  kOtbnAppNotLoaded = 4,
+  /**
+   * The execution of the application on OTBN failed.
+   *
+   * More specific error information can be obtained with
+   * `dif_otbn_get_err_code()`.
+   */
+  kOtbnExecutionFailed = 5,
+} otbn_result_t;
+
+/**
+ * OTBN context structure.
+ *
+ * Use `otbn_init()` to initialize.
+ */
+typedef struct otbn {
+  /**
+   * The OTBN DIF to access the hardware.
+   */
+  dif_otbn_t dif;
+
+  /**
+   * The application loaded or to be loaded into OTBN.
+   */
+  const otbn_app_t *app;
+
+  /**
+   * Is the application loaded into OTBN?
+   */
+  bool app_is_loaded;
```

Comment:
If these members are private, consider marking them \internal (I would also consider dropping the comments altogether, since they don't need do be documented and are kind of self-describing, but up to you).

Reply:
Well, you can interact with them if you'd like to, as long as you know what you're doing. Meaning: there are use cases for reading these struct members, e.g. to get access to the underlying DIF.

Reply:
> Well, you can interact with them if you'd like to, as long as you know what you're doing.

This should be documented. Under what circumstances is it acceptable to reach into this struct? Expecting users that "know what they're doing" to just reach in makes it harder to refactor this type in the future.

Reply:
I don't see what rules I could put there for now. I added a line that `app` is only valid if `app_is_loaded` is true, but beyond that, there's nothing magic about these members. Also, our code base is tiny enough that a quick git grep should make any refactoring to those members a painless job.